### PR TITLE
Enable local clang-tidy lint

### PR DESCRIPTION
**Pull request by 1ntEgr8**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61121**

This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

Differential Revision: [D29598927](https://our.internmc.facebook.com/intern/diff/D29598927)

### CODE REVIEW

File: Makefile
Comment (janeyx99) [inline Makefile:108]: 🎉

File: Makefile
Comment (janeyx99) [inline Makefile:107]: could this be clang-tidy to be consistent with the previous GHA workflow name? 
```suggestion
clang-tidy:
```

**Review by janeyx99 — APPROVED** [review]
nice :D 

don't land before fixing flake8 tho

File: tools/linter/clang_tidy/__main__.py
Comment (driazati) [inline tools/linter/clang_tidy/__main__.py:77]: I'd prefer if we kept `actions_local_runner.py` as a main dispatcher for local lints so we can avoid duplicating stuff like `--changed-only`, it could dispatch out to this script but could you add it as an explicit job here?

https://github.com/pytorch/pytorch/blob/master/tools/actions_local_runner.py#L407

If you do that and implment `quick()`, `full()`, and `filter_files()` this should all come for free. It should work to pass the files list through manually with the `--path` option here so we don't have to add any new ones.

File: Makefile
Comment (driazati) [inline Makefile:108]: This can call the local runner instead which can encode all the args like `--parallel` and the `--clang-tidy-exe`

File: tools/actions_local_runner.py
Comment (driazati) [inline tools/actions_local_runner.py:316]: It looks like this script always returns `0` as its exit code, so even if there are failures the runner still reports that it passed (and it doesn't print the output when the command passes). Can you fix up the clang-tidy script to thread through the return code + unify them from parallel runs?

File: tools/linter/clang_tidy/__main__.py
Comment (driazati) [inline tools/linter/clang_tidy/__main__.py:175]: Can you throw this into the `make setup_lint` command too?

File: tools/linter/clang_tidy/run.py
Comment (driazati) [inline tools/linter/clang_tidy/run.py:283]: nit: unused function?

**Review by driazati — APPROVED** [review]
lgtm, it should be fine once we get the clang-tidy runner to properly report its return code

File: tools/actions_local_runner.py
Comment (1ntEgr8) [reply tools/actions_local_runner.py:316]: `clang-tidy` returns a zero exit code even if there are warnings. To resolve this comment, I'm using the `extract_warnings` method to detect if there are any warnings, and then return the number of warnings as the exit code.

**Comment by 1ntEgr8** [comment]
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29598927).

**Comment by 1ntEgr8** [comment]
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29598927).

**Pull request closed without merging**
